crates/corro-types/src/channel.rs

```diff
@@ -0,0 +1,107 @@
+use metrics::{gauge, Gauge};
+use std::{
+    fmt::{self, Debug, Formatter},
+    time::Duration,
+};
+use tokio::sync::mpsc::{
+    channel,
+    error::{SendError, SendTimeoutError, TryRecvError, TrySendError},
+    Receiver, Sender,
+};
+
+/// A tokio channel wrapper sender that tracks occupancy
+pub struct CorroSender<T> {
+    gauge: Gauge,
+    inner: Sender<T>,
+}
+
+impl<T> Clone for CorroSender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            gauge: self.gauge.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T: Debug> Debug for CorroSender<T> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+/// A tokio channel wrapper receiver that tracks occupancy
+pub struct CorroReceiver<T> {
+    gauge: Gauge,
+    inner: Receiver<T>,
+}
+
+impl<T: Debug> Debug for CorroReceiver<T> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+/// Create a bounded channel which tracks occupancy with a label
+pub fn bounded<T>(max: usize, label: &'static str) -> (CorroSender<T>, CorroReceiver<T>) {
+    let gauge = gauge!("corro.runtime.channel_occupancy", "max_capacity" => max.to_string(), "label" => label);
+    let (tx, rx) = channel(max);
+
+    (
+        CorroSender {
+            gauge: gauge.clone(),
+            inner: tx,
+        },
+        CorroReceiver { gauge, inner: rx },
+    )
+}
+
+impl<T> CorroSender<T> {
+    pub async fn send(&self, value: T) -> Result<(), SendError<T>> {
+        self.inner.send(value).await.map(|r| {
```

Comment:
We might as well track how long it took to send the value into the channel. This would tell us if there's backpressure from channels and how much. This gives us an extra bit of info compared to knowing that a channel does not have capacity anymore.